Pass over the first section on downloading & decoding images (#732)
browserengineering/book: book/embeds.md

    # ...
    while True:
        line = response.readline().decode("utf8")
        # ...
    # ...
```

Note that I _didn't_ add a `decode` call when we read the body; that's
because the body might actually be binary data, and we want to return
that binary data directly to the browser. Now, every existing call to
`request`, which wants textual data, needs to `decode` the response.
For example, in `load`, you'll want to do something like this:

``` {.python replace=Tab/Frame}
class Tab:
    def load(self, url, body=None):
        # ...
        headers, body = request(url, self.url, payload=body)
        body = body.decode("utf8")
        # ...
```

Make sure to make this change everywhere in your browser that you call
`request`, including inside `XMLHttpRequest_send` and in several other

> places in `request`. When we download images, however, we _won't_ call

Reviewer: ```suggestion
places in `load`. When we download images, however, we _won't_ call
```

Reviewer: What about this proposed change?

``` {.python replace=tab/frame}
def download_image(image_src, tab):
    image_url = resolve_url(image_src, tab.url)
    assert tab.allowed_request(image_url), \
        "Blocked load of " + image_url + " due to CSP"
    header, body = request(image_url, tab.url)
    data = skia.Data.MakeWithoutCopy(body)
    img = skia.Image.MakeFromEncoded(data)
    assert img, "Failed to recognize image format for " + image_url
    return body, img
```

Let's look at the steps between `request` and `return` carefully.
First, the requested data is turned into a Skia `Data` object using
the `MakeWithoutCopy` method. Then that `Data` is used to create an
`Image` using `MakeFromEncoded`.

`MakeWithoutCopy` means that the `Data` object just stores a reference
to the existing `body` and doesn't own that data. That's essential,
because encoded image data can be large---maybe megabytes---and
copying that data wastes memory and time. But that means that the
`data` is invalid if `body` is ever garbage-collected, so we return
the `body` from `download_image` and need to make sure to store it
somewhere for at least as long as we're using the image.[^memoryview]

> [^memoryview]: This is a bit of a hack. Perhaps a better solution

Reviewer: I don't quite understand why it's a hack, can you explain?

Author: Uhh, well, IDK, I think a "nice" solution wouldn't involve working around garbage collector weaknesses. In C++ I would "move" the request object into the Skia Data, and or writing directly to a Data would be nicer. But perhaps what you really mean is—let's explain the footnote a bit better so people aren't confused. Will do.

Reviewer: ok sounds good

stored as four bytes, so a 12 megapixel camera (as you can find on
phones these days) produces 48 megabytes of raw data.

[^time-memory]: Decoding costs both a lot of memory and also a lot of
    time, since just writing out all of those bytes can take a big
    chunk of our render budget. More generally, images are more
    expensive than text in a lot of ways. They take a long time to
    download; decoding is slow; resizing is slow; and they can take up
    a lot of both CPU and GPU memory. Optimizing image handling is
    essential to a performant browser.
    
[^html-image-decode]: There's also is an [HTML API][html-image-decode]
    to control decoding, so that the web page author can indicate when
    to pay the cost of decoding.

[html-image-decode]: https://developer.mozilla.org/en-US/docs/Web/API/HTMLImageElement/decoding

But because image decoding can be so expensive, Skia actually has
several algorithms for decoding, some of which result in a
worse-looking image but are faster than the default.[^lossy] For
example, just for resizing an image, there's fast, simple, "nearest
neighbor" resizing and the slower but higher-quality "bilinear" or

> even "[Lanczos][lanczos]" resizing algorithm.

Reviewer: ```suggestion
even "[Lanczos][lanczos]" resizing algorithms.
```

[lanczos]: https://en.wikipedia.org/wiki/Lanczos_resampling

> To give web page authors control over this performance bottleneck, can

Reviewer: ```suggestion
To give web page authors control over this performance bottleneck,
```

there's an [`image-rendering`][image-rendering] CSS property that
indicates which algorithm to use. Let's add that as an argument to
`DrawImage`:

[image-rendering]: https://developer.mozilla.org/en-US/docs/Web/CSS/image-rendering

``` {.python}
class DrawImage(DisplayItem):
    def __init__(self, image, rect, quality):
        # ...
        if quality == "high-quality":
            self.quality = skia.FilterQuality.kHigh_FilterQuality
        elif quality == "crisp-edges":
            self.quality = skia.FilterQuality.kLow_FilterQuality
        else:
            self.quality = skia.FilterQuality.kMedium_FilterQuality

    def execute(self, canvas):
        paint = skia.Paint(FilterQuality=self.quality)
        canvas.drawImageRect(self.image, self.rect, paint)
```

> So we've now got the helper functions we need to downloading and

Reviewer: ```suggestion
So we've now got the helper functions we need to download and
```

> decoding images. But to actually put images into web pages, we're

Reviewer: ```suggestion
decode images. But to actually put images into web pages, we're
```

[^lossy]: Image formats like JPEG are [*lossy*][lossy], meaning that
    they don't faithfully represent all of the information in the
    original picture, so there's a time/quality trade-off going on
    before the file is saved. Typically these formats try to drop
    "noisy details" that a human is unlikely to notice; different
    decoding algorithms are making the same trade-off.

[lossy]: https://en.wikipedia.org/wiki/Lossy_compression

::: {.further}
The HTTP `Content-Type` header lets the web server tell the browser
whether a document contains text or binary data. The header contains a
value called a [MIME type][mime-type]: `text/html`, `text/css`, and
`text/javascript` for HTML, CSS, and JavaScript; `image/png` and
`image/jpeg` for PNG and JPEG images; and [many others][mime-list] for
different font, video, audio, and data formats.[^mime-history]
Interestingly, when we used Skia's `MakeFromEncoded`, we didn't need
to pass in the image format. That's because many image formats start
with ["magic bytes"][magic-bytes]; for example, PNG files always start
with byte 137 followed by the letters "PNG". These magic bytes are

> often more reliable than server-send MIME types, so this kind of

Reviewer: ```suggestion
often more reliable than web-server-provided MIME types, so this kind of
```

[^mime-history]: "MIME" stands for Multipurpose Internet Mail
Extensions, and was originally intended for enumerating all of the
acceptable data formats for email attachments. These days the loop has
basically closed: most email clients are now "webmail" clients,
accessed through your browser, and most emails are now HTML, encoded
with the `text/html` MIME type. Many mail clients do still have an
option to encode the email in `text/plain`, however.

[mime-list]: https://www.iana.org/assignments/media-types/media-types.xhtml

[magic-bytes]: https://www.netspi.com/blog/technical/web-application-penetration-testing/magic-bytes-identifying-common-file-formats-at-a-glance/

Embedded layout
===============

An `<img>` is a leaf element of the DOM. In some ways, it's similar to a single
font glyph that has to paint in a single rectangle sized to the image (instead
of the glyph), takes up space in a `LineLayout`, and causes line breaking when
it reaches the end of the available space.

But it's different than a text *node*, because the text in a text node is not
just one glyph, but an entire run of text of a potentially arbitrary length,

> and that can be split into words and lines across multiple lines. An image, on

Reviewer: ```suggestion
and that can be split into multiple words and lines. An image, on
```

`request`, which wants textual data, needs to `decode` the response.
For example, in `load`, you'll want to do something like this:

``` {.python replace=Tab/Frame}
class Tab:
    def load(self, url, body=None):
        # ...
        headers, body = request(url, self.url, payload=body)
        body = body.decode("utf8")
        # ...
```

Make sure to make this change everywhere in your browser that you call
`request`, including inside `XMLHttpRequest_send` and in several other
places in `request`. When we download images, however, we _won't_ call
`decode`, and just use the binary data directly:

``` {.python replace=tab/frame}
def download_image(image_src, tab):
    image_url = resolve_url(image_src, tab.url)
    if not tab.allowed_request(image_url):
        print("Blocked image", link, "due to CSP")
        return
    try:
        header, body = request(image_url, tab.url)
        return PIL.Image.open(io.BytesIO(body))
    except:
        print("Failed to load image: " + image_url)
```

This method wraps the image bytes as a `BytesIO` object, and then
passes that to `open` to turn it into an Pillow `Image` object.

::: {.todo}

> Chris, do we actually need MIME types? It seems kind of dangerous the

Reviewer: I guess we don't need it after your proposed refactoring.

Reviewer: We could remove it or move it to a go-further block.

Author: I'll make a go-further block out of it.

    return headers, body
```

Interestingly, we don't consult `Content-Type` to decode images.
That's because Pillow auto-detects the image format by peeking at the
first few bytes of the binary data. Image formats typically include
"magic bytes" that identify the image format at the beginning.

:::

> Before we can use a downloaded image, we need to _decode_ it. That's

Reviewer: ```suggestion
After we can use a downloaded image, we need to _decode_ it. That's
```

Author: Don't think the proposal is correct but will rephrase.

Pillow's `tobytes` method lets us convert an encoded Pillow `Image` to
a raw Skia `Image`:

``` {.python replace=(image)/(image%2c%20width%2c%20height%2c%20image_quality)}
def decode_image(image):
    if image.mode != "RGBA":
        image = image.convert("RGBA")
    return skia.Image.frombytes(
        array=image.tobytes(),
        dimensions=image.size,
        colorType=skia.kRGBA_8888_ColorType)
```

This method also converts the image to RGBA order[^rgba-order] before
reading out the raw byte data with `tobytes`.[^maybe-decode]

[^rgba-order]: Which we chose in [Chapter
11](/visual-effects.html#sdl-creates-the-window) as the byte order of
our SDL window.

[^maybe-decode]: In principle, nothing in the Pillow API indicates
_when_ Pillow actually performs the decoding: eagerly when we `open`
the file or lazily when we call `tobytes`. This may seem like an
obscure issue, but in a real browser the timing of a decode is
important for performance. That's also why there is an [HTML
API][html-image-decode] to control decoding.

[html-image-decode]: https://developer.mozilla.org/en-US/docs/Web/API/HTMLImageElement/decoding

::: {.todo}

> I'm a little confused why we use PIL at all, seeing as Skia seems to

Reviewer: I didn't know it had this support. If it works then we can get rid of PIL.

have an [Image.MakeFromEncoded][make-from-encoded] method. Is it just
so we have a chance to talk about the sub-steps like resizing and
resizing algorithms? Is it to support more formats? Skia seems to
have all the common ones:

> Encoded streams supported include BMP, GIF, HEIF, ICO, JPEG, PNG,
> WBMP, WebP.

As a minor point, the Skia API lets us skip creating a `BytesIO`
object or converting things to RGBA.

[make-from-encoded]: https://kyamagu.github.io/skia-python/reference/skia.Image.html#skia.Image.MakeFromEncoded
:::

Because image formats use many clever and sophisticated algorithms,
image decoding is slow and consumes a lot of memory.^[More generally,
images are more expensive than text in a lot of ways. They take a long
time to download; decoding is slow; resizing is slow; and they can
take up a lot of both CPU and GPU memory. Optimizing image handling is

> essential to a performance browser.] Decoding might involve resolving

Reviewer: ```suggestion
essential to a performant browser.] Decoding might involve resolving
```
